packages/fetch-and-compile/lib/fetch.ts

```diff
@@ -0,0 +1,167 @@
+          if (!(error instanceof InvalidNetworkError)) {
+            //if it's *not* just an invalid network, log the error.
+            recognizer.markBadFetcher(Fetcher.fetcherName);
+          }
+          //either way, filter this fetcher out
+          return null;
+        }
+      })
+    )
+  ).filter(fetcher => fetcher !== null);
+  //now: the main loop!
+  let address: string;
+  while ((address = recognizer.getAnUnrecognizedAddress()) !== undefined) {
+    let found: boolean = false;
+    let failureReason: FailureType | undefined; //undefined if no failure
+    //(not set if there is no source)
+    for (const fetcher of fetchers) {
+      //now comes all the hard parts!
+      //get our sources
+      let result: SourceInfo | null;
+      try {
+        debug("getting sources for %s via %s", address, fetcher.fetcherName);
+        result = await fetcher.fetchSourcesForAddress(address);
+      } catch (error) {
+        debug("error in getting sources! %o", error);
+        failureReason = "fetch";
+        continue;
+      }
+      if (result === null) {
+        debug("no sources found");
+        //null means they don't have that address
+        continue;
+      }
+      //if we do have it, extract sources & options
+      debug("got sources!");
+      const { sources, options } = result;
+      if (options.language === "Vyper") {
+        //if it's not Solidity, bail out now
+        debug("found Vyper, bailing out!");
+        recognizer.markUnrecognizable(address);
+        //break out of the fetcher loop, since *no* fetcher will work here
+        break;
+      }
+      //set up the config
+      let externalConfig: Config = config.with({
+        compilers: {
+          solc: options
+        }
+      });
+      //if using docker, transform it (this does nothing if not using docker)
+      externalConfig = transformIfUsingDocker(externalConfig, config);
+      //compile the sources
+      let compileResult: WorkflowCompileResult;
+      try {
+        compileResult = await Compile.sources({
+          options: externalConfig,
+          sources
+        });
+      } catch (error) {
+        debug("compile error: %O", error);
+        failureReason = "compile";
+        continue; //try again with a different fetcher, I guess?
+      }
+      //add it!
+      await recognizer.addCompiledInfo(
+        compileResult,
+        result,
+        address,
+        fetcher.fetcherName
+      );
+      failureReason = undefined; //mark as *not* failed in case a previous fetcher failed
+      //check: did this actually help?
+      debug("checking result");
+      if (!recognizer.getUnrecognizedAddresses().includes(address)) {
+        debug(
+          "address %s successfully recognized via %s",
+          address,
+          fetcher.fetcherName
+        );
+        found = true;
+        //break out of the fetcher loop -- we got what we want
+        break;
+      }
+      debug("address %s still unrecognized", address);
```

Comment:
Sure, I guess I can break this down some more.

Reply:
Hm, it's not really possible to move the body of this loop to a function because of all the `break`s and `continue`s it uses.  Not unless that function were to, like, return a value to indicate whether to break or continue... :-/

Reply:
There's also the way `failureReason` persists until contradicted... that would have to be passed in as an argument to this function...

Reply:
Hm I guess if we made this a little bigger we could turn the inner loop as a whole into a function, rather than just its body?  And then maybe it'd be possible to break things up more?  Not entirely clear though due to the breaks & continues; haven't checked yet.

Reply:
Yeah, let's start with turning the inner loop as a whole into a function

Reply:
OK, did that, hoping that's sufficient...?

Reply:
Thinking about this and discussing it a bit, this is sufficient for now because we'll have opportunity to revisit control flow when we improve error reporting at the interface level.